### ISSUE TITLE
Give it some structure

### ISSUE THREAD

**Issue by AlexanderFabisch** [mlaux1/rl-blox#48]
When we agree on some guidelines, I think we should document them in the readme or in the new sphinx documentation. I just compiled a preliminary API documentation and noticed that we are pretty inconsistent. For new developments (and slowly migrating the existing implementations), we should agree on several things. I will update this when we came to agreements.

I use #49 to explore some useful features that we have to agree on.

## Up for Discussion

- [x] Common NN interface: Do we want to agree on a common NN interface? I'd propose `flax.nnx`. So far I have seen pure JAX, PyTorch, flax.linen and flax.nnx. I think flax.nnx is the most recent API of flax, flax is the best NN interface for JAX and JAX > PyTorch.
- [x] How do we call the main training loop of an algorithm? So far I followed the pattern `train_ALGORITHM(...)`. Unfortunately, I also used `train_ALGORITHM_epoch` in the past if it runs for one epoch. I've also seen `ALGORITHM`. What's the best option?
- [x] What should the main training loop do? Run for one epoch? Run for `total_timesteps` timesteps? Run for one timestep? Run for `total_episodes` episodes? Should it depend on the algorithm? We could also let the user initialize the algorithm state outside of the function call and call the training loop multiple times. This is a bit how `nnx.Optimizer(...)` works.
- [x] Internals: since we want to have everything reusable internal functions and data structures, e.g., losses, networks, etc. should be reusable. I don't know which problems we have and how to improve. I just put the point here and I think we have to look at specific examples.
    - [x] One example: in #49 I created generic MLP and GaussianMLP classes that can be used for $v(s)$, $\pi(a|s)$, and potentially also environment or reward models in the future. For stochastic policies I wrap these two classes to provide additionally the functions `sample(obs)` and `log_probability(obs, act)`.
    - [x] Melvin: `I need some more time to think about how to best organise this.`
- [x] We need a generic logger interface. I played around with one in #49 . You can log any statistics (e.g., losses, average return, episode lengths), but also save checkpoints with orbax. Works pretty nicely so far. What is missing are some evaluation and summary features, logging of experiment configuration, more readable terminal output (maybe use [rich](https://github.com/Textualize/rich)?) and a serialization method. This is inspired by tensorboard or commercial platforms like weights and biases. Maybe we can also reuse a good library? (Maybe not tensorboard.) I think it should also be part of the documentation to point out which models we can checkpoint and which statistics will be logged.
- [x] Algorithm interface arguments: The first argument of the train function is the Gymnasium environment, followed by data containers like function approximators and optimizers. (Hyper-)parameters of the algorithm interface should be of "native" types to enable hyperparameter tuning and experiment tracking. Native types are bool, int, float, tuple, list, dict. A callable is not a native type. At the end, we allow to pass objects like loggers.

## Development Guidelines for RL-BLOX

I will summarize the result of this discussion here.

* We work with `flax.nnx`.
* Two-level API:
    * *high-level* **algorithm level API** that provides a collection of preimplemented algorithms based on a
    * *low-level* **toolbox API** (blox interface) which provides a collection of building blocks to easily create RL algorithms, e.g. value networks, exploration strategies, loss functions, replay buffers, environment models, etc.
* Project structure
    * Related tasks
        * #71 
```
├── examples
├── rl_blox
│ ├── algorithm (contains high level algorithm API)
│ ├── blox (contains low level toolbox API)
│ ├── logging (contains logging with AIM)
│ ├── util (contains all misc utilities, currently in helper and tools)
├── test
└── ...
```
* The **algorithm interface**
    * ... defines a function `train_ALGORITHM` for each algorithm.
    * Argument order
        * `env` - the first argument of the train function is the Gymnasium environment
        * Followed by data containers like function approximators and optimizers.
            * In most algorithms, we currently have the mandatory input of the optimizers. I think these could also be optional and default to a reasonable choice.
        * (Hyper-)parameters of the algorithm interface should be of "native" types to enable hyperparameter tuning and experiment tracking. Native types are bool, int, float, tuple, list, dict. A callable is not a native type.
            * Parameters that are present in almost all algorithms are `total_timesteps` and `seed`.
            * We use the more self-explanatory name `learning_rate` for alpha.
        * `logger` - at the end, we allow to pass objects like loggers, that are not essential for the algorithm, but for the tooling around training
    * The main training loop of an algorithm should run for `total_timesteps` as a default termination condition as this will ensure that algorithms will have the specified number of interactions with the environment, which might not be the case in envs with variable episode lengths otherwise and number of interactions is a key metric when evaluating sample efficiency. If algorithms state a specific termination condition, then this should probably the one we use in the algorithm level API.
    * It would be good to have clearly defined default values for all optional inputs, ideally to have the default behaviour be as close as possible to the original paper, e.g., using Rprop as an optimizer in DQN. The default values should be clearly documented and easy to find for users.
    * Related tasks
        * https://github.com/mlaux1/rl-blox/issues/62
        * https://github.com/mlaux1/rl-blox/issues/92
* The **blox interface**
    * Related tasks
        * #66
        * #68
* Full experiment evaluation and tracking, hyperparameter tuning should not be something this library is meant to provide and remain separate.
    * We provide some basic logging and experiment tracking with AIM.
    * We need to ensure that rl-blox can easily be use in libraries that provide these features, e.g. optuna or hyperopt.
    * Related tasks
        * [ ] We miss some evaluation and summary features, logging of experiment configuration, more readable terminal output (maybe use [rich](https://github.com/Textualize/rich)?) and a serialization method.

**Comment by mlaux1** [comment]
I agree that we need to make some structural decisions and document them. Here are my thoughts on the points you've raised:

- Common NN interface: I agree we should work with flax.nnx, I find it more intuitive than linen
- Name of main training loop: Consistent naming would be good. I have no strong opinion here. `ALGORITHM` would be the most logical to me as we are using an algorithm to train an agent, but `train_ALGORITHM` provides a nice recognisable keyword in the function name. Not a big fan of adding `epoch` as it does not seem to consistently have the same meaning across algorithms and domains.
- Main training loop: Good question. I think there are good arguments for both `total_timesteps` and `total_episodes`, but am leaning towards `total_timesteps` as a default termination condition as this will ensure that algorithms will have the specified number of interactions with the environment, which might not be the case in envs with variable episode lengths otherwise and number of interactions is a key metric when evaluating sample efficiency. If algorithms state a specific termination condition, then this should probably the one we use in the algorithm level API
- Internals: Agreed, this is what I had in mind when I stated the 3 design principles in the readme. I think we should go for a two-level API: one algorithm level API that provides a collection of preimplemented algorithms based on a lower lever toolbox API which provides a collection of building blocks to easily create RL algorithms, e.g. value networks, exploration strategies, loss functions, replay buffers, environment models, etc... I need some more time to think about how to best organise this.
-  Logger: Yes, I like the lightweight wrapper you implemented in the linked PR. I think we should go for some basic logging and experiment tracking with aim, but full experiment evaluation and tracking, hyperparameter tuning should not be something this library is meant to provide and remain separate. We need to ensure that rl-blox can easily be use in libraries that provide these features, e.g. optuna or hyperopt.

**Comment by mlaux1** [comment]
├── examples
├── rl_blox
│   ├── algorithm (contains high level algorithm API)
│   ├── blocks (contains low level toolbox API)
│   ├── logging (contains logging with AIM)
│   ├── util (contains all misc utilities, currently in helper and tools)
├── test
└── ...

@AlexanderFabisch This is my suggestion for organising the project internally, what do you think? I currently don't see a strong reason to split the algorithms between model-free and model-based, a flatter hierarchy should be preferred IMO.

**Comment by AlexanderFabisch** [comment]
Looks good.

**Comment by AlexanderFabisch** [comment]
Ah, can we rename `blocks` to `blox`?

**Comment by AlexanderFabisch** [comment]
I collected a bit the results of our discussion and connected it to issues and tasks at the top of this issue.

**Comment by AlexanderFabisch** [comment]
Can we agree on algorithm interface arguments?

The first argument of the train function is the Gymnasium environment, followed by data containers like function approximators and optimizers. (Hyper-)parameters of the algorithm interface should be of "native" types to enable hyperparameter tuning and experiment tracking. Native types are bool, int, float, tuple, list, dict. A callable is not a native type. At the end, we allow to pass objects like loggers.

Some examples:

```python
def train_ac(
    env: gym.Env,
    policy: StochasticPolicyBase,
    policy_optimizer: nnx.Optimizer,
    value_function: MLP,
    value_function_optimizer: nnx.Optimizer,
    policy_gradient_steps: int = 1,
    value_gradient_steps: int = 1,
    total_timesteps: int = 1_000_000,
    gamma: float = 1.0,
    steps_per_update: int = 1_000,
    train_after_episode: bool = False,
    key: jnp.ndarray | None = None,
    logger: LoggerBase | None = None,
):

def train_ddpg(
    env: gym.Env[gym.spaces.Box, gym.spaces.Box],
    policy: nnx.Module,
    policy_optimizer: nnx.Optimizer,
    q: nnx.Module,
    q_optimizer: nnx.Optimizer,
    seed: int = 1,
    total_timesteps: int = 1_000_000,
    buffer_size: int = 1_000_000,
    gamma: float = 0.99,
    tau: float = 0.005,
    batch_size: int = 256,
    gradient_steps: int = 1,
    exploration_noise: float = 0.1,
    learning_starts: int = 25_000,
    policy_frequency: int = 2,
    policy_target: nnx.Optimizer | None = None,
    q_target: nnx.Optimizer | None = None,
    logger: LoggerBase | None = None,
) -> tuple[
    nnx.Module, nnx.Module, nnx.Optimizer, nnx.Module, nnx.Module, nnx.Optimizer
]:

def train_pets(
    env: gym.Env,
    reward_model: Callable[[ArrayLike, ArrayLike], jnp.ndarray],
    dynamics_model: EnsembleTrainState,
    plan_horizon: int,
    n_particles: int,
    n_samples: int,
    n_opt_iter: int = 5,
    init_with_previous_plan: bool = True,
    seed: int = 1,
    buffer_size: int = 1_000_000,
    total_timesteps: int = 1_000_000,
    learning_starts: int = 100,
    learning_starts_gradient_steps: int = 100,
    n_steps_per_iteration: int = 100,
    gradient_steps: int = 10,
    logger: LoggerBase | None = None,
) -> tuple[
    PETSMPCConfig,
    PETSMPCState,
    Callable[
        [GaussianMLPEnsemble, jnp.ndarray, jnp.ndarray, jnp.ndarray],
        jnp.ndarray,
    ],
]:

def train_reinforce(
    env: gym.Env,
    policy: StochasticPolicyBase,
    policy_optimizer: nnx.Optimizer,
    value_function: MLP | None = None,
    value_function_optimizer: nnx.Optimizer | None = None,
    policy_gradient_steps: int = 1,
    value_gradient_steps: int = 1,
    total_timesteps: int = 1_000_000,
    gamma: float = 1.0,
    steps_per_update: int = 1_000,
    train_after_episode: bool = False,
    key: jnp.ndarray | None = None,
    logger: LoggerBase | None = None,
):

def train_sac(
    env: gym.Env[gym.spaces.Box, gym.spaces.Box],
    policy: StochasticPolicyBase,
    policy_optimizer: nnx.Optimizer,
    q1: nnx.Module,
    q1_optimizer: nnx.Optimizer,
    q2: nnx.Module,
    q2_optimizer: nnx.Optimizer,
    seed: int = 1,
    total_timesteps: int = 1_000_000,
    buffer_size: int = 1_000_000,
    gamma: float = 0.99,
    tau: float = 0.005,
    batch_size: int = 256,
    learning_starts: float = 5_000,
    entropy_learning_rate: float = 1e-3,
    policy_frequency: int = 2,
    target_network_frequency: int = 1,
    alpha: float = 0.2,
    autotune: bool = True,
    q1_target: nnx.Module | None = None,
    q2_target: nnx.Module | None = None,
    entropy_control: EntropyControl | None = None,
    logger: LoggerBase | None = None,
) -> tuple[
    nnx.Module,
    nnx.Optimizer,
    nnx.Module,
    nnx.Module,
    nnx.Optimizer,
    nnx.Module,
    nnx.Module,
    nnx.Optimizer,
    EntropyControl,
]:
```

**Comment by AlexanderFabisch** [comment]
@mlaux1 ?

**Comment by mlaux1** [comment]
This is a good point. I was thinking about this while reviewing all the corresponding PRs. I think we should try to standardise this as much as possible, but I think forcing it too much is not the right move at this point. So parameters that will for sure be present in each algorithm, like the ones you mentioned should be standardised, but many will only be present in some algorithms and not in others.

Another thing we should discuss are default values: I think it would be good to have clearly defined default values for all optional inputs, ideally to have the default behaviour be a as close as possible to the original paper, e.g., using Rprop as a loss function in DQN. The default values should be clearly documented and easy to find for users.

Finally, in most algorithms, we currently have the mandatory input of the optimizers. I think these could also be optional and default to a reasonable choice. The algorithm API should ideally be simple to use without the user having to first create complicated structures that even differ from one algorithm to the next. We might need to discuss where we want to draw the line between mandatory algorithm inputs and optional ones that default to reasonable choices.
